IOAPIC intrinsics and both phys and virt classes

connojd opened the pull request:
The register space of the IOAPIC is described in Chapter 3 of
the IOAPIC specification. Its registers may be accessed with 32-bit
loads and stores offset from the base address derived from the ACPI
tables (which defaults to 0xFEC00000).

- Add phys_ioapic class to interact with the physical IOAPIC
- Add virt_ioapic class to interact with virtual IOAPICs
- Add corresponding unit tests

The current implementation provides only basic read/write capabilities
A future iteration is planned to provide more complex logic
(e.g. virtual routing) as needed.

Signed-off-by: Connor Davis <davisc@ainfosec.com>

connojd commented:
@JaredWright You may notice that the ioapic intrinsics namespace is 'ioapic' as opposed to 'ioapic_register' similar to lapic_register. IMO this reads better and is more in-line with the manual. If you like this, I propose we move the contents of lapic_register into lapic.h in hypervisor, and then in another PR I will move ioapic.h from here to hypervisor.

I coult move ioapic.h to hypervisor before this PR is merged; that may make more sense

connojd commented on bfvmm/src/hve/arch/intel_x64/virt_ioapic.cpp:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+//virt_ioapic::virt_ioapic(gsl::not_null<eapis::intel_x64::phys_ioapic *> phys) :
```
oops I forgot these bits

rianquinn commented:
When using std::array, you should be using .at() and not []. 

On a side note, on the next PR, I would like to see Clang Tidy turned back on as it would be catching those types of issues.

JaredWright commented:
Is there a reason that the lapic and x2apic intrinsics are split up between the base the extended API's already? I don't think I see any disadvantages of moving all the intrinsics to the base, because then they would all be in one file together rather than split across repos.

And also, is this [sketchy looking document](https://pdos.csail.mit.edu/6.828/2016/readings/ia32/ioapic.pdf) the IOAPIC specification? I didn't even realize there was a separate spec.

connojd commented:
@rianquinn I already check the bounds of the offset, so at() isn't really needed
@JaredWright thats the one! I'm fine with moving all the apic intrinsics to the base

rianquinn commented:
@connojd Yes. The use of [] is strictly discouraged by the Core Guidelines and will fail on static analysis. If clang tidy were being used, it would fail. 

In fact, Bjorne has said multiple times, hardware has gotten good enough that we can afford the check each time.

connojd commented:
@JaredWright @rianquinn I changed to at(), and I moved the intrinsic code to bfintrinsics (pending PR). Clang tidy appears to be working, at least the output looks very similar to the hypervisor repo.

connojd commented:
closing this until the base intrinsics PR gets merged
